[docs] Convert migration guide to asciidoc (#82600)
elastic/kibana: docs/developer/best-practices/performance.asciidoc

[[plugin-performance]]
== Keep Kibana fast

*tl;dr*: Load as much code lazily as possible. Everyone loves snappy
applications with responsive UI and hates spinners. Users deserve the
best user experiences regardless of whether they run Kibana locally or
in the cloud, regardless of their hardware & environment.

> There are 2 main aspects of the perceived speed of an application: loading time
> and responsiveness to user actions. Kibana Platform loads and bootstraps *all*
> the plugins whenever a user lands on any page. It means that adding
> every new application affects overall *loading performance*, as plugin code is
> loaded *eagerly* to initialize the plugin and provide plugin API to dependent
> plugins.

Reviewer: ```suggestion
There are 2 main aspects of the perceived speed of an application: loading time
and responsiveness to user actions. Kibana loads and bootstraps *all*
the plugins whenever a user lands on any page. It means that
every new application affects the overall *loading performance*, as plugin code is
loaded *eagerly* to initialize the plugin and provide plugin API to dependent
plugins.
```

require UI root components lazily when possible (such as in mount
handlers). Even if their size may seem negligible, they are likely using
some heavy-weight libraries that will also be removed from the initial
plugin bundle, therefore, reducing its size by a significant amount.

[source,typescript]
----
import type { Plugin, CoreSetup, AppMountParameters } from 'kibana/public';
export class MyPlugin implements Plugin<MyPluginSetup> {
  setup(core: CoreSetup, plugins: SetupDeps) {
    core.application.register({
      id: 'app',
      title: 'My app',
      async mount(params: AppMountParameters) {
        const { mountApp } = await import('./app/mount_app');
        return mountApp(await core.getStartServices(), params);
      },
    });
    plugins.management.sections.section.kibana.registerApp({
      id: 'app',
      title: 'My app',
      order: 1,
      async mount(params) {
        const { mountManagementSection } = await import('./app/mount_management_section');
        return mountManagementSection(coreSetup, params);
      },
    });
    return {
      doSomething() {},
    };
  }
}
----

=== Understanding plugin bundle size

> Kibana Platform plugins are distributed as a pre-built with
> `@kbn/optimizer` package artifacts. It allows us to get rid of the
> shipping of `optimizer` in the distributable version of Kibana. Every NP
> plugin artifact contains all plugin dependencies required to run the
> plugin, except some stateful dependencies shared across plugin bundles
> via `@kbn/ui-shared-deps`. It means that NP plugin artifacts tend to
> have a bigger size than the legacy platform version. To understand the
> current size of your plugin artifact, run `@kbn/optimizer` as

Reviewer: ```suggestion
Kibana Platform plugins are pre-built with `@kbn/optimizer` 
and distributed as package artifacts. This means that it is no
longer necessary for us to include the `optimizer` in the 
distributable version of {kib}. Every plugin artifact contains all
plugin dependencies required to run the plugin, except some
stateful dependencies shared across plugin bundles via 
`@kbn/ui-shared-deps`. This means that plugin artifacts tend to
be larger than they were in the legacy platform. To understand the
current size of your plugin artifact, run `@kbn/optimizer` with
```

[source,bash]
----
node scripts/build_kibana_platform_plugins.js --dist --profile --focus=my_plugin    
----

and check the output in the `target` sub-folder of your plugin folder

[source,bash]
----
ls -lh plugins/my_plugin/target/public/
# output
# an async chunk loaded on demand
... 262K 0.plugin.js
# eagerly loaded chunk
... 50K  my_plugin.plugin.js
----

> you might see at least one js bundle - `my_plugin.plugin.js`. This is
> the only artifact loaded by the platform during bootstrap in the
> browser. The rule of thumb is to keep its size as small as possible.
> Other lazily loaded parts of your plugin present in the same folder as
> separate chunks under `{number}.plugin.js` names. If you want to
> investigate what your plugin bundle consists of you need to run
> `@kbn/optimizer` with `--profile` flag to get generated
> https://webpack.js.org/api/stats/[webpack stats file].

Reviewer: ```suggestion
You might see at least one js bundle - `my_plugin.plugin.js`. This is
the only artifact loaded by {kib} during bootstrap in the
browser. The rule of thumb is to keep its size as small as possible.
Other lazily loaded parts of your plugin will be present in the same folder as
separate chunks under `{number}.myplugin.js` names. If you want to
investigate what your plugin bundle consists of you need to run
`@kbn/optimizer` with `--profile` flag to generate a
https://webpack.js.org/api/stats/[webpack stats file].
```

> [[plugin-performance]]

Reviewer: Once we've published these docs it would be nice to link to it from the CI Reporter so that a developer can come here from their PR to read how to improve the bundle size

== Keep Kibana fast

*tl;dr*: Load as much code lazily as possible. Everyone loves snappy
applications with responsive UI and hates spinners. Users deserve the
best user experiences regardless of whether they run Kibana locally or
in the cloud, regardless of their hardware & environment.

There are 2 main aspects of the perceived speed of an application: loading time
and responsiveness to user actions. Kibana loads and bootstraps *all*
the plugins whenever a user lands on any page. It means that
every new application affects the overall _loading performance_, as plugin code is

> loaded _eagerly_ to initialize the plugin and provide plugin API to dependent

Reviewer 2: in this context: `eagerly === immediately`, as a contrast to `lazily === on demand`.

plugins.

However, it’s usually not necessary that the whole plugin code should be loaded
and initialized at once. The plugin could keep on loading code covering API functionality
on Kibana bootstrap but load UI related code lazily on-demand, when an
application page or management section is mounted.
Always prefer to

> require UI root components lazily when possible (such as in `mount`

Reviewer 2: A plugin might require a module in runtime, but it doesn't control the `loading` mechanism -  Kibana does.
```js
const module = await import('./my_module'); <-- A plugin wants to get 'my_module' and knows it might take time to load the bundle, but it doesn't know how Kibana loads one.
```

> require UI root components lazily when possible (such as in `mount`

Reviewer: maybe we can use "import" instead of "require"
